### PR TITLE
Improve measure computer API

### PR DESCRIPTION
**Pull request by julienlancelot**
- Javadoc has been added
- Check of implementation of MeasureComputerDefinition
- Prevent 2 computers to generate the same metric

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/FeedMeasureComputers.java [hunk 1]
```diff
@@ -216,12 +224,22 @@ public boolean apply(@Nullable String metric) {
...
-      if (!pluginMetricKeys.contains(metric)) {
-        throw new IllegalStateException(String.format("Metric '%s' cannot be used as an output metric as no plugin declare this metric", metric));
+    public boolean apply(@Nonnull String metric) {
+      checkState(!CORE_METRIC_KEYS.contains(metric), String.format("Metric '%s' cannot be used as an output metric as it's a core metric", metric));
+      checkState(pluginMetricKeys.contains(metric), String.format("Metric '%s' cannot be used as an output metric as no plugin declare this metric", metric));
+      return true;
+    }
+  }
+
+  private class ValidateUniqueOutputMetric implements Predicate<MeasureComputerWrapper> {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/FeedMeasureComputers.java]: MAJOR Make this a "static" inner class. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2694)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues:
* MAJOR 6 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol/input' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol/input) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol/output' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol/output) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/db/protobuf' package (org.codehaus.sonar:sonar-db:src/main/gen-java/org/sonar/db/protobuf) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonarqube/ws' package (org.codehaus.sonar:sonar-ws:src/main/gen-java/org/sonarqube/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

**Comment by simonbrandhof** [comment]
LGTM

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol/input' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol/input) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/batch/protocol/output' package (org.codehaus.sonar:sonar-batch-protocol:src/main/gen-java/org/sonar/batch/protocol/output) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonar/db/protobuf' package (org.codehaus.sonar:sonar-db:src/main/gen-java/org/sonar/db/protobuf) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/gen-java/org/sonarqube/ws' package (org.codehaus.sonar:sonar-ws:src/main/gen-java/org/sonarqube/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

**Pull request closed without merging**